src/vcpu.rs

```diff
@@ -0,0 +1,204 @@
+// Copyright 2018-2019 CrowdStrike, Inc.
+// SPDX-License-Identifier: Apache-2.0 or MIT
+//
+// Portions Copyright 2018 Cloudbase Solutions Srl
+// SPDX-License-Identifier: Apache-2.0 or MIT
+//
+// Portions Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+//
+// Portions Copyright 2017 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the THIRD-PARTY file.
+
+//! Trait to access the functionality of a virtual CPU (vCPU).
+
+use std::{io, result};
+
+#[cfg(any(target_arch = "arm", target_arch = "aarch64"))]
+pub use arm::VcpuInit;
+
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+pub use x86_64::{
+    StandardRegisters, SpecialRegisters, FpuState, MsrEntries, MsrEntry, 
+    CpuId, LapicState
+};
+
+///
+/// Reasons for vCPU exits.
+///
+/// Exit reasons are drawn a superset of exit reasons for various VMMs.
+///
+#[derive(Debug)]
+pub enum VcpuExit<'a> {
+    None,
```

Comment:
Are the following exits Hyper-v exits? Is there any overlap between KVM and Hyper-v exits? If not I would rather have them as 2 defines: one VcpuExit for Hyper-v and one for KVM. Could you please point me to where are these VM exits defined for Windows? I kinda got lost googling for them.

Reply:
Related to both this comment and the one above: There is some overlap. The top block are Windows-specific, the middle block are KVM-specific, and the bottom block are common between the two. Seeing as there's no real reason to separate them, maybe it makes sense to mix them all together in alphabetical order, as you've suggested (optionally commenting which ones are found in each hypervisor).  An alternative is to keep them subcategorized, add a comment for each block as to where they can be found, and alphabatize within the block. Do you have a preference? 

Documentation for Windows VM exits can be found here: https://docs.microsoft.com/en-us/virtualization/api/hypervisor-platform/funcs/whvrunvirtualprocessor

Reply:
Thanks for the documentation link. Regarding the documentation, we need to have documentation for all entries. For the KVM entries you can get away with copying the documentation from kvm-ioctls.

Regarding keeping them in one enum, I believe the overlap is so small that it is rather confusing to have them as one enum. We can wait for other people to chime in because it might be my personal preference.

Reply:
Yup, adding `#[deny(missing_docs)]`, as you requested above, is indeed revealing many entries for which I am adding documentation, these included :)

I am not opposed to platform-specific VcpuEntries, wrapped with conditional compilation (as discussed in the first round of comments in this PR, an associated type would have been my preference, but I think the requirement of carrying lifetime annotations everywhere makes this unappealing).  Vcpu-defined per-platform exits may be the cleanest solution.

Reply:
Added #[deny(missing_docs)] and updated documentation everywhere to fix failures

Reply:
And, agreeing with your suggestion, I split the VcpuExits up by platform. For each, I reordered Exits by their enum/#define as specified in their relevant header files (they actually were pretty close to this).